thehive/app/org/thp/thehive/services/PatternSrv.scala

```diff
@@ -0,0 +1,81 @@
+package org.thp.thehive.services
+
+import org.apache.tinkerpop.gremlin.structure.Graph
+import org.thp.scalligraph.EntityIdOrName
+import org.thp.scalligraph.auth.AuthContext
+import org.thp.scalligraph.models.{Database, Entity}
+import org.thp.scalligraph.services._
+import org.thp.scalligraph.traversal.TraversalOps._
+import org.thp.scalligraph.traversal.{Converter, Traversal}
+import org.thp.thehive.models._
+import org.thp.thehive.services.CaseOps._
+import org.thp.thehive.services.PatternOps._
+import org.thp.thehive.services.ProcedureOps._
+
+import java.util.{Map => JMap}
+import javax.inject.{Inject, Named, Singleton}
+import scala.util.{Success, Try}
+
+@Singleton
+class PatternSrv @Inject() (
+    auditSrv: AuditSrv,
+    caseSrv: CaseSrv,
+    organisationSrv: OrganisationSrv
+)(implicit
+    @Named("with-thehive-schema") db: Database
+) extends VertexSrv[Pattern] {
+  val patternPatternSrv = new EdgeSrv[PatternPattern, Pattern, Pattern]
+
+  def parentExists(child: Pattern with Entity, parent: Pattern with Entity)(implicit graph: Graph): Boolean =
+    child._id == parent._id || get(child).parent.getEntity(parent).exists
+
+  def setParent(child: Pattern with Entity, parent: Pattern with Entity)(implicit authContext: AuthContext, graph: Graph): Try[Unit] =
+    if (parentExists(child, parent)) Success(())
+    else patternPatternSrv.create(PatternPattern(), parent, child).map(_ => ())
+
+  override def getByName(name: String)(implicit graph: Graph): Traversal.V[Pattern] =
+    Try(startTraversal.getByPatternId(name)).getOrElse(startTraversal.limit(0))
+
+  def getCasePatterns(caseId: String)(implicit graph: Graph): Try[Seq[String]] =
+    for {
+      caze <- caseSrv.get(EntityIdOrName(caseId)).getOrFail("Case")
+      patterns = caseSrv.get(caze).procedure.pattern.richPattern.toSeq
+    } yield patterns.map(_.patternId)
+
+  def remove(pattern: Pattern with Entity)(implicit graph: Graph, authContext: AuthContext): Try[Unit] =
+    for {
+      organisation <- organisationSrv.getOrFail(authContext.organisation)
+      _            <- auditSrv.pattern.delete(pattern, organisation)
+    } yield get(pattern).remove()
+
+}
+
+object PatternOps {
+  implicit class PatternOpsDefs(traversal: Traversal.V[Pattern]) {
+    def get(idOrName: EntityIdOrName): Traversal.V[Pattern] =
+      idOrName.fold(traversal.getByIds(_), _ => traversal.limit(0))
```

Comment:
The pattern name is the `PatternId`, right ?
```suggestion
      idOrName.fold(traversal.getByIds(_), getByPatternId(_))
```